### PR TITLE
[KBFS-1097] Split BlockServerLocal

### PR DESCRIPTION
**Pull request by akalin-keybase**
...into BlockServerMemory and BlockServerDisk.

The latter will eventually turn into a
journaling implementation of BlockServer.

Make both BlockServerMemory and BlockServerDisk
more careful in checking their arguments.

Also make BlockContext into a struct type
embedded into BlockPointer.

Also add a way to use BlockServerDisk in tests,
and do that in CI. (This'll suffice for testing
BlockServerDisk for now.)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -39,11 +39,15 @@ script:
     [ -z "$lint" -o "$lint" = "Lint-free!" ]
   - go vet $(go list ./... 2>/dev/null | grep -v /vendor/)
   - go install github.com/keybase/kbfs/...
-  - cd libkbfs && go test -i && go test -race -c && ./libkbfs.test -test.timeout 4m
+  - cd libkbfs && go test -i && go test -race -c && ./libkbfs.test -test.timeout 1m
```
Comment (akalin-keybase) [inline .travis.yml]: Bumped down the timeout here

File: libfuse/mount_test.go [hunk 2]
```diff
@@ -2305,11 +2305,13 @@ func (t *testMountObserver) TlfHandleChange(ctx context.Context,
 func TestInvalidateAcrossMounts(t *testing.T) {
 	config1 := libkbfs.MakeTestConfigOrBust(t, "user1",
 		"user2")
+	defer libkbfs.CheckConfigAndShutdown(t, config1)
```
Comment (akalin-keybase) [inline libfuse/mount_test.go]: These were leaking configs

File: appveyor.yml [hunk 3]
```diff
@@ -66,6 +66,7 @@ build_script:
   - echo github.com/keybase/kbfs/libkbfs > testlist.txt
   - echo github.com/keybase/kbfs/libdokan >> testlist.txt
   - echo github.com/keybase/kbfs/test >> testlist.txt
+# Run the below tests with env KEYBASE_TEST_BSERVER_ADDR=tempdir.
```
Comment (strib) [inline appveyor.yml]: Should this be prefixed with "TODO:"?

File: libkbfs/kbfs_ops_test.go [hunk 4]
```diff
@@ -574,8 +574,11 @@ func makeBP(id BlockID, rmd *RootMetadata, config Config,
 		ID:      id,
 		KeyGen:  rmd.LatestKeyGeneration(),
 		DataVer: config.DataVersion(),
-		Creator: u,
-		// refnonces not needed for tests until dedup is implemented
+		BlockContext: BlockContext{
+			Creator: u,
+			// refnonces not needed for tests until dedup
```
Comment (strib) [inline libkbfs/kbfs_ops_test.go]: Ouch, this comment is really old.  While you're here, do you mind changing it to something like "Explicit refnonce testing happens elsewhere"?

File: libkbfs/bserver_disk.go [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/keybase/client/go/logger"
+	"golang.org/x/net/context"
+)
+
+// BlockServerDisk implements the BlockServer interface by just
+// storing blocks in a local leveldb instance
```
Comment (strib) [inline libkbfs/bserver_disk.go]: Missing trailing period.

Comment (strib) [inline libkbfs/bserver_disk.go]: Several more below.  (I don't actually care much, but I think I've seen you point it out before so I thought you might want to know.)

File: libkbfs/bserver_memory.go [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+		config.Crypto(),
+		config.MakeLogger("BSM"),
+		sync.RWMutex{},
+		make(map[BlockID]blockMemEntry),
+	}
+}
+
+var errBlockServerMemoryShutdown = errors.New("BlockServerMemory is shutdown")
+
+// Get implements the BlockServer interface for BlockServerMemory
```
Comment (strib) [inline libkbfs/bserver_memory.go]: More period-less comments in this file too.

File: libkbfs/bserver_memory.go [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+	}
+
+	if entry, ok := b.m[id]; ok {
+		if entry.tlfID != tlfID {
+			return fmt.Errorf(
+				"TLF ID mismatch: expected %s, got %s",
+				entry.tlfID, tlfID)
+		}
+
+		// TODO: Is it okay for the creator to change?
```
Comment (strib) [inline libkbfs/bserver_memory.go]: I don't think so.  Puts should really only ever happen once, though we should handle duplicate Puts idempotently.

File: libkbfs/bserver_memory.go [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+		if refEntry, ok := entry.refs[zeroBlockRefNonce]; ok && refEntry.Context != context {
+			return fmt.Errorf("Context mismatch: expected %s, got %s",
+				refEntry.Context, context)
+		}
+
+		// The only thing that could be different now is the
+		// key server half.
+		//
+		// TODO: Figure out whether it can actually be
```
Comment (strib) [inline libkbfs/bserver_memory.go]: Shouldn't ever change, since that would change the block ID, right?

File: libkbfs/bserver_memory.go [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+		count, err := b.removeBlockReferences(id, tlfID, idContexts)
+		if err != nil {
+			return nil, err
+		}
+		liveCounts[id] = count
+	}
+	return liveCounts, nil
+}
+
+// ArchiveBlockReferences implements the BlockServer interface for
```
Comment (strib) [inline libkbfs/bserver_memory.go]: Wrong function comment.

File: libkbfs/bserver_tlf_storage.go [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+// TODO: Have bserverTlfStorage store a journal of put operations.
+
+// TODO: Do all high-level operations atomically on the file-system
+// level.
+
+// bserverTlfStorage stores block data for a single TLF in flat files
+// on disk. In particular, it stores data for each block in a
+// directory with name equal to the hex-encoded blockID.
+//
+// The block ID name is splayed over 256^2 subdirectories (one byte
```
Comment (strib) [inline libkbfs/bserver_tlf_storage.go]: Given that there's unlikely to ever be 256 supported hashes, this number seems a bit large and confused me at first.  Maybe reword to make it clearer that currently there are only 256 possible directories?

File: libkbfs/bserver_tlf_storage.go [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+	return res, nil
+}
+
+func (s *bserverTlfStorage) putRefEntryLocked(
+	id BlockID, refEntry blockRefEntry) error {
+	buf, err := s.codec.Encode(refEntry)
+	if err != nil {
+		return err
+	}
+	// TODO: Add integrity-checking?
```
Comment (strib) [inline libkbfs/bserver_tlf_storage.go]: Yeah, it's weird that the memory server checks but this one doesn't.  Can probably factor some of it out into a common helper.

Comment (strib) [inline libkbfs/bserver_tlf_storage.go]: Oh wait, the caller already does a `validateBlockServerPut`.  Nevermind, I misread.

File: libkbfs/test_common.go [hunk 12]
```diff
@@ -88,17 +89,22 @@ func MakeTestConfigOrBust(t logger.TestLogBackend,
...
 	bserverAddr := os.Getenv(EnvTestBServerAddr)
-	if len(bserverAddr) != 0 {
-		blockServer :=
-			NewBlockServerRemote(config, bserverAddr)
-		config.SetBlockServer(blockServer)
-	} else {
-		blockServer, err := NewBlockServerMemory(config)
+	var blockServer BlockServer
+	switch {
+	case bserverAddr == "tempdir":
```
Comment (strib) [inline libkbfs/test_common.go]: "tempdir" would be clearer as a constant.

File: .travis.yml [hunk 13]
```diff
@@ -39,11 +39,15 @@ script:
...
+  - env KEYBASE_TEST_BSERVER_ADDR=tempdir ./libkbfs.test -test.timeout 1m
   - cd ../libfuse && go test -i && go test -c && ./libfuse.test -test.timeout 2m
+  - env KEYBASE_TEST_BSERVER_ADDR=tempdir ./libfuse.test -test.timeout 2m
   - cd ../test
   - go test -i -tags fuse && go test -race -c && ./test.test -test.timeout 4m
+  - env KEYBASE_TEST_BSERVER_ADDR=tempdir ./test.test -test.timeout 4m
   - go test -c -tags fuse && ./test.test -test.timeout 4m
+  - env KEYBASE_TEST_BSERVER_ADDR=tempdir go test -c -tags fuse && ./test.test -test.timeout 4m
```
Comment (strib) [inline .travis.yml]: Overall, the new tests seems to have added about 80 seconds to the build.  I think it's probably overkill.  Specifically in this directory I don't think we need to run both of the libkbfs- and the fuse-based tests; the way they interact with the bserver should be exactly the same.  Can we eliminate one?

**Comment by strib** [comment]
Looks good to me, mod nits.  Thanks!

File: appveyor.yml (hunk 3 above)
Comment (akalin-keybase) [inline appveyor.yml]: Done

File: libkbfs/kbfs_ops_test.go (hunk 4 above)
Comment (akalin-keybase) [inline libkbfs/kbfs_ops_test.go]: Done

File: libkbfs/bserver_disk.go (hunk 5 above)
Comment (akalin-keybase) [inline libkbfs/bserver_disk.go]: Done

File: libkbfs/bserver_memory.go (hunk 6 above)
Comment (akalin-keybase) [inline libkbfs/bserver_memory.go]: Done

File: libkbfs/bserver_memory.go (hunk 7 above)
Comment (akalin-keybase) [inline libkbfs/bserver_memory.go]: Okay, added some more checking.

File: libkbfs/bserver_memory.go (hunk 8 above)
Comment (akalin-keybase) [inline libkbfs/bserver_memory.go]: You're right -- added more checking.

File: libkbfs/bserver_memory.go (hunk 9 above)
Comment (akalin-keybase) [inline libkbfs/bserver_memory.go]: Fixed.

File: libkbfs/bserver_tlf_storage.go (hunk 10 above)
Comment (akalin-keybase) [inline libkbfs/bserver_tlf_storage.go]: Rewrote comment to be hopefully clearer.

File: libkbfs/test_common.go (hunk 12 above)
Comment (akalin-keybase) [inline libkbfs/test_common.go]: Done.

File: .travis.yml (hunk 13 above)
Comment (akalin-keybase) [inline .travis.yml]: As discussed, do only in-memory bserver for libkbfs, and on-disk bserver for everything else.
